### DISCUSSION TITLE
Individual device cost statistics

### DISCUSSION THREAD

**Discussion by TillFleisch** [home-assistant/architecture#790] [General]
## Background
The `energy` integration supports cost calculation and statistics for power consumption from the grid. The price is either directly provided by an entity or calculated using a fixed/fluctuating price.

This cost-tracking feature is not available for individual devices, but has been requested [here](https://community.home-assistant.io/t/energy-cost-for-the-individual-devices/334932).
Cost statistics cannot be calculated, since no settings are known for individual devices.
This raises the question: How can we calculate the usage cost for each device individually.

## Proposal
The two obvious solutions:

1. Copy&Pase settings and calculation from `grid consumption`
2. Add individual devices to a parent grid (grid consumption device) and use the grid's cost settings to determine the cost for each device

**Pros and Cons of approach 1**
Pros:
- Easy to implement
- flexible in terms of different individual device costs

Cons:
- makes setting up the energy dashboard complex and bloated
- adds redundant information (devices likely have the same cost settings as grid consumption)
- we do not know to which grid a device is connected

**Pros and Cons of approach 2**

Pros:
- less bloated energy dashboard setup
- changing the energy price only requires changing the parents' price
- we know which device is connected to which grid

Cons:
- Cannot represent different cost for a specific device (unlikely)
- Does not work without a parent

In my opinion, solution 2 is better than solution 1.
The consumption of individual devices is probably also measured by the grid consumption meter. Thus, they should have the same energy cost settings. Using these settings for individual devices  should yield correct pricing.

In these Pull Requests ([core](https://github.com/home-assistant/core/pull/76925), [frontend](https://github.com/home-assistant/frontend/pull/13396)) i propose a implementation which uses approach 2 to provide individual device cost statistics.

This approach also allows the following (which approach 1 does not):
Since we know how devices are connected, we can create a graph showing where energy is going over time. We can subsidize the measured grid consumption with the per device consumption.

Very scientific example (would probably look better with more devices i.e. heatpump):
The green parts represent the partial consumption measured from the individual device 'car charger'.
![image](https://user-images.githubusercontent.com/20563797/185176175-f0961bdc-8c36-4df5-b406-ba09fefc21a4.png)
*(example) combined consumption graph*

In an ideal world, each device would have a power meter and the sum of all devices should add up to whatever the grid consumption meter measures.

## Questions
- Different approaches for determining the cost of individual devices
  - possibly combine 1 and 2 into a better solution?
- Further flaws of approach 2
- Which approach is more user-friendly
- How can we realize the combined consumption graph using other methods (i.e. approach 1)

**Comment by dgomes** [comment]
I think this feature makes lots of sense, specially if you have power hungry devices such as EVSE or Heatpumps for which tracking cost is very advisable.

Approach 2 also makes more sense IMHO

**Comment by frenck** [comment]
To me is not clear what "parent grid" means at all. Maybe that is a Dutch thing, that is possible, but we have a single power grid in NL, so selecting one seems odd?

Do you have different outlets connected to different grids or something? And then decide from which grid you consume energy by plugging in the device in either of the outlets?

**Reply by TillFleisch** [reply]
I presume someone could have a separate grid for an EV-Charger or a Workshop/Shed.

**Reply by frenck** [reply]
Right, but its still the same grid :)

So I talked to Paulus (who responded down below, too), and I learned indeed that those are groups/networks of power, possibly with additional meters; which is more common in other countries (not NL 😄). Home Assistant doesn't have that concept at this point, it assumes/supports a single grid at this point.

The ability to add multiple grid sources/entities to the energy dashboard for the grid, is based on things like tariffs from a single grid, not multiple grids.

**Reply by TillFleisch** [reply]
> The ability to add multiple grid sources/entities to the energy dashboard for the grid, is based on things like tariffs from a single grid, not multiple grids.

I only learned this today. In my case I used the three different phases from a power meter as separate entities. This gives some information about where power is used. (i.e. different floors can be connected to different phases)

**Reply by frenck** [reply]
Hehe well, at least its clear why I was confused about "parent grid" 😄

**Comment by balloob** [comment]
Right now Home Assistant assumes there is only 1 grid. Each consumption, return to grid, solar and battery is assumed to be connected to the same net.

In countries like Germany, one can be connected to multiple electricity networks, one for EV charging and one for the rest. If we want to allocate devices to grids, we can't just link it to a "parent" consumption source. If there are 2 consumption sources representing 2 tariffs, it already doesn't work. It also doesn't take solar/battery energy into account that might have been used.

So to do it correct, we would first need to introduce the notion of "networks" into HA energy management. Each network can have a completely different configuration of input/output/solar and it's linked devices.

**Reply by TillFleisch** [reply]
I agree the above presented approaches can only determine the theoretical usage cost of a device. If someone wanted to optimize a device to use more solar power this wouldn't be very helpful.

Perhaps the introduction of Groups could solve this Problem. A group being a collection of input, output, solar, battery and individual devices. Essentially just having multiple instances of the current energy management.  

This only works if those implicitly assumed networks are independent of one another.
(I don't know if this is usually the case)

This would also require multiple Energy Dashboards and possibly one which combines everything into one overview.

To me this seems like too much of a niche Case(for now).

Maybe the solution to this problem is the introduction of a theoretical cost section.
This section would include the energy cost without solar/battery compensation(what you would pay without solar power). Such statistics could help determine cost savings from solar power.
(This could include the cost, as proposed above, for individual devices)
